### PR TITLE
V2 control transfer target device

### LINKED ISSUE #58: Setting UVCC autoWhiteBalance to 0 Getting LIBUSB_ERROR_IO
`sudo uvcc devices`

```
Optical Mouse [ vId: 0x458  / pId: 0x3a  ]
USB Keyboard [ vId: 0x4d9  / pId: 0x1702  ]
EHCI Host Controller [ vId: 0x1d6b  / pId: 0x2  ]
EHCI Host Controller [ vId: 0x1d6b  / pId: 0x2  ]
USB 2.0 Camera [ vId: 0xc45  / pId: 0x6366  ]
```

`sudo uvcc --vendor 0xc45 --product 0x6366 set autoWhiteBalance 0`

`Error: LIBUSB_ERROR_IO`

Screenshot from 2019-07-24 17-18-41

### LINKED ISSUE #60: Example : dwExposureTimeAbsolute is not working
https://github.com/makenai/node-uvc-control/blob/master/lib/controls.js#L92

Because the Type of `EXPOSURE_TIME_ABSOLUTE_CONTROL`(`CT_EXPOSURE_TIME_ABSOLUTE_CONTROL = 0x04` ) is PU for some reason, if you try to change `CT_EXPOSURE_TIME_ABSOLUTE_CONTROL`, the gain of PU (`PU_GAIN_CONTROL = 0x04`) will be changed.

AS-IS: Gain of PU is changed.
TO-BE: Exposure_Time_Absolute is changed.

### LINKED ISSUE #62: Logitech C270 manual exposure
Hi,

I downloaded the repository, ran `npm install` then `node test/discover.js` but there's no output. I'm testing the Logitech C270 on MacOS 10.14.6, with the goal of manually adjusting exposure for this webcam.

When I run `npm run preview` I can select the Logitech C270 webcam from the dropdown, it shows up as: `USB Camera (046d:0825)`.

Most of the time, changing the sliders won't affect anything.

I get this error:
```
setting absolute_exposure_time [ '72' ]
{ Error: LIBUSB_ERROR_NO_DEVICE
    at Device.usb.Device.controlTransfer (~/Downloads/node-uvc-control-master/node_modules/usb/usb.js:136:12)
    at Promise (~/Downloads/node-uvc-control-master/index.js:227:19)
    at new Promise (<anonymous>)
    at UVCControl.set (~/Downloads/node-uvc-control-master/index.js:219:12)
    at app.post (~/Downloads/node-uvc-control-master/examples/server.js:43:7)
    at Layer.handle [as handle_request] (~/Downloads/node-uvc-control-master/node_modules/express/lib/router/layer.js:95:5)
    at next (~/Downloads/node-uvc-control-master/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (~/Downloads/node-uvc-control-master/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (~/Downloads/node-uvc-control-master/node_modules/express/lib/router/layer.js:95:5)
    at ~/Downloads/node-uvc-control-master/node_modules/express/lib/router/index.js:281:22 errno: -4 }
```
Turns out I need to wait for the webcam to show up in Chrome, before I run `npm run preview`. Then I don't get the above error.

By default, the value for "bAutoExposureMode" is set to "MANUAL". However, in reality this is auto exposure mode. When I set it to "AUTO", then it's still in an auto exposure mode, but now the image is exposed somewhat darker... There's also "PRIORITY_SHUTTER" and "PRIORITY_APERTURE". These are also auto exposure modes... None of these modes disable auto exposure and there's no way to get back to the initial default (brighter) auto exposure, unless I unplug the webcam. When I open the Logitech Camera Settings app and hit "Restore Defaults", the image becomes brighter again, but now auto exposure is actually disabled (which it isn't by default!).

Alternatively I can open [CameraControl](https://phoboslab.org/log/2009/07/uvc-camera-control-for-mac-os-x) and uncheck "Auto Exposure", to enable manual exposure. This also allows me to re-enable auto exposure (this checkbox works perfectly fine! :) However, the "Exposure Time" slider in CameraControl doesn't work well. Most of the slider range doesn't change exposure. Some spots in the range make the image slightly darker. Moving the slider all the way to the right makes the image super dark, under-exposing so much everything becomes black except for lamps themselves.
Screenshot 2020-02-21 at 01 22 21
This is pretty cool and using the controls from the node-uvc-control web interface don't allow me to under expose this much. With gain and brightness all the way to the left I still see myself sitting in the room. So what does CameraControl do differently that it can under expose so much?

My issue is that exposure control for the C270 webcam doesn't work with node-uvc-control, but it should be possible because CameraControl can enable manual exposure control (and under expose even more). Would it be possible to support this model and what can I do to help?

### PR DESCRIPTION
**Pull request by joelpurra**
Work-in-progress patches to the future `node-uvc-control` v2, on `master`. Fixes Linux issues (primarily) but also compatibility with other cameras (confirmed by some `uvcc` users). Sorry for mixing issues in one pull request -- submitting this as a draft pull request for testing, and I'd be happy to split to multiple when it's confirmed to work.

- Fixes #58.
- Fixes #60.
- Fixes #62.

Fixes were created to get `uvcc` working, so testing (unfortunately still manually) is mostly done using the `uvcc` CLI.

- https://github.com/joelpurra/uvcc/issues/4

While `uvcc` v2 is not up to par with `node-uvc-control` v2, I've also fixed a few more basic `uvcc` features to get basic `uvcc export | uvcc import` functionality working. The `feature/uvc-control-v2-next` branch of `uvcc` targets the code in this pull request.

- https://github.com/joelpurra/node-uvc-control/tree/v2-control-transfer-target-device
- https://github.com/joelpurra/uvcc/tree/feature/uvc-control-v2-next

```shell
git clone "git@github.com:joelpurra/uvcc.git"
cd uvcc

git checkout feature/uvc-control-v2-next
git pull
rm -r node_modules/
npm install

./index.js --verbose ranges
./index.js --verbose export

# NOTE: not all settings can be set if they're in auto-mode -- fix pending.
./index.js --verbose export | ./index.js --verbose import
```

There are still bugs, both in `uvcc` (missing functionality, setting values when they're in auto-mode, etcetera) and `node-uvc-control` (spotted some on the byte-reading level). Tested on macOS and Ubuntu using a Logitech C920 (0x46d/0x82d) with node v12.

What are the results for your camera(s)? Any major issues? Please show some output for verification, along with the camera vendor/product ids.

---

Note: similar/backported fixes for Linux are available for `node-uvc-control` v1.0.2 in a separate branch, but github doesn't let me create pull request against tags.

- https://github.com/makenai/node-uvc-control/issues/58#issuecomment-540815829

### CODE REVIEW

**Comment by nickrobillard** [comment]
Hi. I've tried to use this but I am running into an error. The file `${__dirname}/../node_modules/uvc-control/list-devices.js` is included in `uvcc/src/usb-device-lister.js` but it's missing. Looks like you were changing a file directly in your `node_modules` dir and may not have added it back to the repo?

Here's what I did:
```
git clone "git@github.com:joelpurra/uvcc.git"
cd uvcc

git checkout feature/uvc-control-v2-next
git pull
rm -r node_modules/
npm install

./index.js --verbose devices
```
Error occurs:
```
Parsed arguments: {
  "_": [],
  "version": false,
  "verbose": true,
  "help": false,
  "$0": "index.js",
  "cmd": "devices"
}
internal/modules/cjs/loader.js:985
  throw err;
  ^

Error: Cannot find module '/[redacted]/uvcc/node_modules/uvc-control/list-devices.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)
    at Function.Module._load (internal/modules/cjs/loader.js:864:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12)
    at internal/main/run_main_module.js:18:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

**Comment by joelpurra** [comment]
@nickrobillard: I skipped device listing for this preview. It was using a file from the "old" `uvc-control` v1 which is no longer there.

- Reimplemented `uvcc devices`.
- Made the device address configurable, for those who have more than one camera of the same make.
- Made device configuration unnecessary for the most common use-case -- having a single UVC camera connected. The device vendor/product ids and address can be auto-detected by `uvc-control` v2 (it picks the "first" camera), so `uvcc` no longer requires these values to be configured.

Basically `uvcc devices` is no longer as important because of the new auto-detection, and `uvcc` usage is greatly simplified.

---

Note: all changes were made in `uvcc` v2, so they are not visible in this pull request.

- https://github.com/joelpurra/uvcc/tree/feature/uvc-control-v2-next

You can install and test the new code as described in the original pull request information.

**Comment by joelpurra** [comment]
Closing in favor of #66.
